File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

// Write your first then apply function here
export function firstThenApply<T, U>(
  arr: T[],
  predicate: (x: T) => boolean,
  transform: (x: T) => U
): U | undefined {
  const found = arr.find(predicate)
  return found === undefined ? undefined : transform(found)
}

// Write your powers generator here
export function* powersGenerator(base: bigint): Generator<bigint> {
  let power: bigint = 1n
  while (true) {
    yield power
    power *= base
  }
}

// Write your line count function here
export async function meaningfulLineCount(filename: string): Promise<number> {
  try {
    const file = await open(filename)
    let count = 0
    
    for await (const line of file.readLines()) {
      const trimmedLine = line.trim()
      if (trimmedLine && !trimmedLine.startsWith('#')) {
        count++
      }
    }
    
    return count
  } catch (error: any) {
    throw new Error(`Error reading file: ${error.message}`)
  }
}

// Write your shape type and associated functions here
export type Shape = 
  | { kind: "Sphere"; radius: number }
  | { kind: "Box"; width: number; length: number; depth: number }

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return (4/3) * Math.PI * Math.pow(shape.radius, 3)
    case "Box":
      return shape.width * shape.length * shape.depth
  }
}

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return 4 * Math.PI * Math.pow(shape.radius, 2)
    case "Box":
      return 2 * (
        shape.width * shape.length +
        shape.width * shape.depth +
        shape.length * shape.depth
      )
  }
}

// Write your binary search tree implementation here
export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0
  }

  contains(_value: T): boolean {
    return false
  }

  insert(value: T): BinarySearchTree<T> {
    return new Node(value, new Empty(), new Empty())
  }

  *inorder(): Generator<T> {
    // Empty generator - yields nothing
  }

  toString(): string {
    return "()"
  }
}

class Node<T> implements BinarySearchTree<T> {
  constructor(
    private readonly value: T,
    private readonly left: BinarySearchTree<T>,
    private readonly right: BinarySearchTree<T>
  ) {}

  size(): number {
    return 1 + this.left.size() + this.right.size()
  }

  contains(value: T): boolean {
    if (value === this.value) return true
    if (value < this.value) return this.left.contains(value)
    return this.right.contains(value)
  }

  insert(value: T): BinarySearchTree<T> {
    if (value === this.value) return this
    if (value < this.value) {
      return new Node(this.value, this.left.insert(value), this.right)
    }
    return new Node(this.value, this.left, this.right.insert(value))
  }

  *inorder(): Generator<T> {
    yield* this.left.inorder()
    yield this.value
    yield* this.right.inorder()
  }

  toString(): string {
    const leftStr = this.left.toString()
    const rightStr = this.right.toString()
    if (leftStr === "()" && rightStr === "()") {
      return `(${this.value})`
    }
    const left = leftStr === "()" ? "" : leftStr
    const right = rightStr === "()" ? "" : rightStr
    return `(${left}${this.value}${right})`
  }
}

export interface BinarySearchTree<T> {
  size(): number
  contains(value: T): boolean
  insert(value: T): BinarySearchTree<T>
  inorder(): Generator<T>
  toString(): string
}

File: haskell/Exercises.hs
module Exercises
    ( change,
      firstThenApply,
      powers,
      meaningfulLineCount,
      Shape(..),
      volume,
      surfaceArea,
      BST(Empty),
      size,
      contains,
      insert,
      inorder
    ) where

import qualified Data.Map as Map
import Data.Text (pack, unpack, replace)
import Data.List(isPrefixOf, find)
import Data.Char(isSpace)

change :: Integer -> Either String (Map.Map Integer Integer)
change amount
    | amount < 0 = Left "amount cannot be negative"
    | otherwise = Right $ changeHelper [25, 10, 5, 1] amount Map.empty
        where
          changeHelper [] remaining counts = counts
          changeHelper (d:ds) remaining counts =
            changeHelper ds newRemaining newCounts
              where
                (count, newRemaining) = remaining `divMod` d
                newCounts = Map.insert d count counts

firstThenApply :: [a] -> (a -> Bool) -> (a -> b) -> Maybe b
firstThenApply xs p f = f <$> find p xs

powers :: Integral a => a -> [a]
powers base = iterate (base *) 1

meaningfulLineCount :: FilePath -> IO Int
meaningfulLineCount filename = do
    contents <- readFile filename
    return $ length $ filter isNotCommentOrEmpty $ lines contents
    where
        isNotCommentOrEmpty line = 
            let trimmed = dropWhile isSpace line
            in not (null trimmed) && head trimmed /= '#'

data Shape = Sphere Double | Box Double Double Double
    deriving (Show, Eq)

volume :: Shape -> Double
volume (Sphere r) = (4/3) * pi * r^3
volume (Box w l d) = w * l * d

surfaceArea :: Shape -> Double
surfaceArea (Sphere r) = 4 * pi * r^2
surfaceArea (Box w l d) = 2 * (w*l + w*d + l*d)

data BST a = Empty | Node a (BST a) (BST a)
    deriving (Eq)

instance (Show a, Eq a) => Show (BST a) where
    show Empty = "()"
    show (Node v Empty Empty) = "(" ++ show v ++ ")"
    show (Node v l r) = "(" ++ leftStr ++ show v ++ rightStr ++ ")"
        where
            leftStr = if l == Empty then "" else show l
            rightStr = if r == Empty then "" else show r

size :: BST a -> Int
size Empty = 0
size (Node _ l r) = 1 + size l + size r

contains :: Ord a => a -> BST a -> Bool
contains _ Empty = False
contains x (Node v l r)
    | x == v = True
    | x < v = contains x l
    | otherwise = contains x r

insert :: Ord a => a -> BST a -> BST a
insert x Empty = Node x Empty Empty
insert x t@(Node v l r)
    | x == v = t
    | x < v = Node v (insert x l) r
    | otherwise = Node v l (insert x r)

inorder :: BST a -> [a]
inorder Empty = []
inorder (Node v l r) = inorder l ++ [v] ++ inorder r
